File: autopal-express/src/metrics.js
const counters = new Map();

function increment(metric, amount = 1) {
  if (amount < 0) {
    throw new Error('amount must be non-negative');
  }
  if (amount === 0) {
    return;
  }
  counters.set(metric, (counters.get(metric) || 0) + amount);
}

function snapshot() {
  return Object.fromEntries(counters);
}

module.exports = {
  increment,
  snapshot
};

File: autopal/app/metrics.py
"""Lightweight in-process counters for operational telemetry."""

from __future__ import annotations

from collections import Counter
from typing import Dict

_COUNTERS: Counter[str] = Counter()


def increment(metric: str, *, amount: int = 1) -> None:
    """Increment ``metric`` by ``amount`` (default: 1)."""

    if amount < 0:
        raise ValueError("amount must be non-negative")
    if amount:
        _COUNTERS[metric] += amount


def snapshot() -> Dict[str, int]:
    """Return a copy of the current counter values."""

    return dict(_COUNTERS)


__all__ = ["increment", "snapshot"]
